ocdata/core/adsorbate_slab_config.py

```diff
@@ -0,0 +1,339 @@
+                triangle_positions = surface_atoms_pos[tri]
+                sites = get_random_sites_on_triangle(
+                    triangle_positions, num_sites_per_triangle
+                )
+                all_sites += sites
+            np.random.shuffle(all_sites)
+        elif self.mode == "heuristic":
+            ###
+            # Comment(@abhshkdz): Initial attempt at not making the trip from
+            # Ase to Pymatgen and back. But there are differences in the sites
+            # returned by the two methods. So, we'll stick to Pymatgen for now.
+            #
+            # https://pymatgen.org/pymatgen.analysis.adsorption.html
+            # Pymatgen assigns on-top, bridge, and hollow adsorption sites at
+            # the nodes, edges, and face centers of the Delaunay triangulation.
+            # https://github.com/materialsproject/pymatgen/blob/v2023.3.23/pymatgen/analysis/adsorption.py#L217
+            # ontop: sites on top of surface atoms.
+            # all_sites += [site for site in surface_atoms_pos]
+            # for tri in simplices:
+            #     # bridge: sites at centers of edges between surface atoms.
+            #     for vpair in [[0, 1], [1, 2], [2, 0]]:
+            #         all_sites.append(
+            #             np.mean(surface_atoms_pos[tri][vpair], axis=0)
+            #         )
+            #     # hollow: sites at centers of Delaunay triangulation faces.
+            #     # Note that we currently don't implement the check to avoid
+            #     # sampling hollow sites in obtuse triangles:
+            #     # https://github.com/materialsproject/pymatgen/blob/v2023.3.23/pymatgen/analysis/adsorption.py#L272
+            #     all_sites.append(
+            #         np.mean(surface_atoms_pos[tri], axis=0)
+            #     )
+            ###
+            # Explicitly specify "surface" / "subsurface" atoms so that
+            # Pymatgen doesn't recompute tags.
+            site_properties = {"surface_properties": []}
+            for atom in self.slab.atoms:
+                if atom.tag == 1:
+                    site_properties["surface_properties"].append("surface")
+                else:
+                    site_properties["surface_properties"].append("subsurface")
+            struct = AseAtomsAdaptor.get_structure(self.slab.atoms)
+            # Copy because Pymatgen doesn't let us update site_properties.
+            struct = struct.copy(site_properties=site_properties)
+            asf = AdsorbateSiteFinder(struct)
+            # `distance` refers to the distance along the surface normal between
+            # the slab and the adsorbate. We set it to 0 here since we later
+            # explicitly check for atomic overlap and set the adsorbate height.
+            all_sites += asf.find_adsorption_sites(distance=0)["all"]
+            np.random.shuffle(all_sites)
+        else:
+            raise NotImplementedError
+
+        return all_sites[:num_sites]
```

Comment:
heuristic should not be restricted to num_sites, only random

Reply:
Updated: https://github.com/Open-Catalyst-Project/Open-Catalyst-Dataset/blob/refactor-v2/ocdata/core/adsorbate_slab_config.py#L149-L155